soc: stm32: add SWO logger backend support

mrfuchs opened the pull request:
This patch adds HAS_SWO selections to all STM32 SoCs supporting Serial
Wire Output via the Trace Port Interface Unit (TPIU).

Signed-off-by: Markus Fuchs <markus.fuchs@de.sauter-bc.com>

mrfuchs committed:
soc: stm32: add SWO logger backend support

This patch adds HAS_SWO selections to all STM32 SoCs supporting Serial
Wire Output via the Trace Port Interface Unit (TPIU).

Signed-off-by: Markus Fuchs <markus.fuchs@de.sauter-bc.com>

arnop2 commented on soc/arm/st_stm32/stm32h7/Kconfig.series:
```diff
@@ -14,6 +14,7 @@ config SOC_SERIES_STM32H7X
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
 	select NEWLIB_LIBC
 	select USE_STM32_HAL_RCC_EX if CPU_CORTEX_M4
+	select HAS_SWO
```
Please could you  crosscheck that SWO backend  you want to enable, is compatible with all the stm32 SoC?  
Regarding the SWO backend implementation, the expected path is ITM->TPIU->TRACESWO.
I took a look to the STM32F4 datasheet, it is ok. But regarding  stm32H7 and stm32MP1, the backend has changed, probably due to multi processor architecture...
At least for these two SoCs the TPIU no longer seems used but replaced by the SWO block to output on TRACESWO.
So not compatible with the current SWO backend

mrfuchs replied:
You are right. Looking at the debug infrastructure block diagram for the STM32H742 (figure 816 in RM0433) and comparing it to the one for the STM32F437 (figure 485 in RM0090) shows that the TPIU is no longer routed to TRACESWO. I missed that. As you said, this works for the STM32F4 - that's the SoC I've used for testing - but it won't for some others in this patch. I'll re-examine and update the patch. Thanks for spotting this.

mrfuchs replied:
I double-checked the reference manuals and reverted the changes related to the STM32H7, STM32MP1 and STM32WB SoCs, as they do not support SWO via the TPIU.

erwango commented:
Looks good to me, but you'll need to follow carefully following change: https://github.com/zephyrproject-rtos/zephyr/pull/16958/files#diff-de8bc5ba726c0901ca1b703ad923f307

ioannisg commented:
@mrfuchs could you address the feedback from @erwango ?

erwango reviewed (changes requested):
Following merge of #16958, SWO subsystem uses the following definition:
```
#define SWO_FREQ (DT_CPU_CLOCK_FREQUENCY + (CONFIG_LOG_BACKEND_SWO_FREQ_HZ / 2))
```
`DT_CPU_CLOCK_FREQUENCY` is not generated for STM32 based boards.
Sop either it is added, either this definition should be amended.

mrfuchs commented:
@erwango So, due to #16958 this PR now kind of depends on #17748, right?

erwango commented:
@mrfuchs 

or SWO code should be changed not to use `DT_CPU_CLOCK_FREQUENCY`.

arnop2 commented:
> @mrfuchs
> 
> or SWO code should be changed not to use `DT_CPU_CLOCK_FREQUENCY`.

FYI, Arm coresight also propose a SWO component  which is not part of the CPU (TPIU is not used). In this case the SWO output frequency is not derivate from the CPU freq but from the SWO coresight component input clock.
Still need to figure out how to integrate this new coresight component, but could help to have a more generic name used...

mrfuchs commented:
> or SWO code should be changed not to use `DT_CPU_CLOCK_FREQUENCY`.

Yes, that should work as well. However, I feel like this should be done in a separate PR then, as this patch only changes configuration. It can be applied to the 1.14 branch and it'll work out of the box as with my STM32F4 + ST-Link setup.

erwango commented:
@mrfuchs, 

> Yes, that should work as well. However, I feel like this should be done in a separate PR then, as this patch only changes configuration. It can be applied to the 1.14 branch and it'll work out of the box as with my STM32F4 + ST-Link setup.

Not fully sure if this kind of enhancement could be back-ported, but I think it would acceptable.
Anyway, ok to create a new PR for that.

Also, STM32WB and STM32L1 also supports SWO, can you update them as well.

mrfuchs commented:
> Also, STM32WB and STM32L1 also supports SWO, can you update them as well.

Thanks for your feedback. I added the STM32L1 - must have missed its compatibility. As for the STM32WB, I re-added it as well. I removed it with the last change referring to the DBG block diagram for CPU2 (which is a Cortex-M0+ and not compatible with the current SWO backend implementation). However, CPU1 (which is a Cortex-M4) indeed supports SWO via TPIU according to the reference manual.
